### PR TITLE
[11.x] Improve Precognition

### PR DESCRIPTION
**Pull request by mehdi-fathi**
It seems the code `CanBePrecognitive` trait needed a method to get `Precognition-Validate-Only` header data in a proper way. It would be useful for debugging purposes. In addition, `PrecognitionTest` file doesn't have some asserts with `Precognition` header in some actions.

### CODE REVIEW

File: src/Illuminate/Http/Concerns/CanBePrecognitive.php [hunk 1]
```diff
@@ -28,7 +28,7 @@ public function filterPrecognitiveRules($rules)
      *
      * @return bool
      */
-    public function isAttemptingPrecognition()
+    public function isAttemptingPrecognition(): bool
```
Comment (timacdonald) [inline src/Illuminate/Http/Concerns/CanBePrecognitive.php:31]:
```suggestion
    public function isAttemptingPrecognition()
```

File: src/Illuminate/Http/Concerns/CanBePrecognitive.php [hunk 2]
```diff
@@ -12,14 +12,14 @@ trait CanBePrecognitive
      * @param  array  $rules
      * @return array
      */
-    public function filterPrecognitiveRules($rules)
+    public function filterPrecognitiveRules(array $rules): array
```
Comment (timacdonald) [inline src/Illuminate/Http/Concerns/CanBePrecognitive.php:15]:
```suggestion
    public function filterPrecognitiveRules($rules)
```

**Comment by timacdonald** [comment]
I don't mind extracting a method, but could you not do your debugging in the `filterPrecognitiveRules` method?

```php
public function filterPrecognitiveRules($array)
{
    $rules = parent::filterPrecognitiveRules($array);

    if ($this->isPrecognitive() {
        // debugging...
    }

    return $rules;
}
```

File: src/Illuminate/Http/Concerns/CanBePrecognitive.php (hunk 1 above)
Comment (mehdi-fathi) [reply src/Illuminate/Http/Concerns/CanBePrecognitive.php:31]: @timacdonald Thanks for checking the merge.

I thought return-type declarations would enhance the performance although it's generally minimal.

**Comment by mehdi-fathi** [comment]
@timacdonald 

Yes, we can debug there but a developer might need a method to access the prepared array data. 

`filterPrecognitiveRules` is able to provide data with filtered by rules. If a developer wants to see 'Precognition-Validate-Only' as a prepared array, they should use the code `explode(',', $this->header('Precognition-Validate-Only'))`.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**